README.rst
===============================
Click utils
===============================

.. image:: https://badge.fury.io/py/click-utils.png
    :target: http://badge.fury.io/py/click-utils

.. image:: https://travis-ci.org/slafs/click-utils.png?branch=master
        :target: https://travis-ci.org/slafs/click-utils

.. image:: https://pypip.in/d/click-utils/badge.png
        :target: https://pypi.python.org/pypi/click-utils


a set of utilites for writing command line programs with Click

* Free software: BSD license
* Documentation: https://click-utils.readthedocs.org.

Features
--------

* Couple of useful ``logging`` options and types
* A custom command and option class for printing env vars names (WIP)
* More to come...


HISTORY.rst
.. :changelog:

History
-------

0.2.0 (2015-01-30)
---------------------

* modify ``loglevel_option`` (is_eager, callback with storing value on context)
* add an abstract ``logger_option`` and a ``logfile_option``

0.1.0 (2015-01-26)
---------------------

* First release on PyPI.
* initial project structure
* Add LogLevelChoice type
* Add loglevel_option
* Add logconfig_option


tests/test_help.py
# -*- coding: utf-8 -*-

import click

import click_utils


def test_help_env(runner):

    @click.command()
    def cli(option_a, option_b, option_c, option_d):
        click.echo(option_a)
        click.echo(option_b)
        click.echo(option_c)
        click.echo(option_d)

    option = click_utils.help.EnvHelpOption(['--option_a'], envvar='OPTIONA')
    cli.params.append(option)
    option = click_utils.help.EnvHelpOption(['--option_b'], envvar=['BVAR', 'B_VAR'])
    cli.params.append(option)
    option = click_utils.help.EnvHelpOption(['--option_c'])
    cli.params.append(option)
    option = click_utils.help.EnvHelpOption(['--option_d'], allow_from_autoenv=False)
    cli.params.append(option)

    result = runner.invoke(cli, ['--help'], catch_exceptions=False, auto_envvar_prefix='CLI')
    out = result.output
    assert '--option_a TEXT  [env: OPTIONA]' in out
    assert '--option_b TEXT  [env: BVAR, B_VAR]' in out
    assert '--option_c TEXT  [env: CLI_OPTION_C]' in out
    assert '--option_d TEXT  [env: ' not in out

    # result = runner.invoke(cli,  ['--option_a=A', '--option_b=B'], catch_exceptions=False)


def test_help_command(runner):
    @click.command(cls=click_utils.EnvHelpCommand)
    @click.option('--option_a', envvar='OPTIONA')
    @click.option('--option_b', envvar=['BVAR', 'B_VAR'])
    @click.option('--option_c', )
    @click.option('--option_d', allow_from_autoenv=False)
    def cli(option_a, option_b, option_c, option_d):
        click.echo(option_a)
        click.echo(option_b)
        click.echo(option_c)
        click.echo(option_d)

    result = runner.invoke(cli, ['--help'], catch_exceptions=False, auto_envvar_prefix='CLI')
    out = result.output
    assert '--option_a TEXT  [env: OPTIONA]' in out
    assert '--option_b TEXT  [env: BVAR, B_VAR]' in out
    assert '--option_c TEXT  [env: CLI_OPTION_C]' in out
    assert '--option_d TEXT  [env: ' not in out


tests/test_types.py
# -*- coding: utf-8 -*-

import logging
import click

import click_utils


def test_loglevel_type(runner):
    '''
    test a typical case with deafult log levels
    '''

    @click.command()
    @click.option('--loglevel', type=click_utils.LogLevelChoice())
    def cli(loglevel):
        click.echo('loglevel: {0}'.format(loglevel))

    result = runner.invoke(cli,  ['--loglevel=error'], catch_exceptions=False)

    assert result.output == 'loglevel: 40\n'

    result = runner.invoke(cli,  ['--help'])

    assert '--loglevel [notset|debug|info|warning|error|critical]' in result.output


def test_loglevel_type_extra(runner):
    '''
    test a case when a user specifies a custom logging level
    '''

    logging.addLevelName(102, 'My test level')

    @click.command()
    @click.option('--loglevel', type=click_utils.LogLevelChoice(101, 102))
    def cli(loglevel):
        click.echo('loglevel: {0}'.format(loglevel))

    result = runner.invoke(cli,  ['--help'])

    assert '--loglevel [notset|debug|info|warning|error|critical|level101|mytestlevel]' in result.output

    result = runner.invoke(cli,  ['--loglevel=mytestlevel'], catch_exceptions=False)

    assert result.output == 'loglevel: 102\n'


def test_loglevel_numeric_input(runner):
    '''
    test a special case where user inputted a number loglevel instead of a level name
    '''
    @click.command()
    @click.option('--loglevel', type=click_utils.LogLevelChoice())
    def cli(loglevel):
        assert isinstance(loglevel, int)
        click.echo('loglevel: {0}'.format(loglevel))

    result = runner.invoke(cli,  ['--loglevel=123'], catch_exceptions=False)

    assert result.output == 'loglevel: 123\n'


def test_loglevel_wrong_input(runner):
    '''
    test invokation with wrong input
    '''
    @click.command()
    @click.option('--loglevel', type=click_utils.LogLevelChoice())
    def cli(loglevel):
        click.echo('loglevel: {0}'.format(loglevel))

    result = runner.invoke(cli,  ['--loglevel=nonexistentlevel'], catch_exceptions=True)

    assert result.exit_code != 0
    assert 'Invalid value for "--loglevel": invalid choice: nonexistentlevel' in result.output


tests/test_logging.py
# -*- coding: utf-8 -*-

import logging
import click
import textwrap

import click_utils


def test_loglevel_type(runner):
    '''
    test a typical case with deafult log levels
    '''

    @click.command()
    @click.option('--loglevel', type=click_utils.LogLevelChoice())
    def cli(loglevel):
        click.echo('loglevel: {0}'.format(loglevel))

    result = runner.invoke(cli,  ['--loglevel=error'], catch_exceptions=False)

    assert result.output == 'loglevel: 40\n'

    result = runner.invoke(cli,  ['--help'])

    assert '--loglevel [notset|debug|info|warning|error|critical]' in result.output


def test_loglevel_type_extra(runner):
    '''
    test a case when a user specifies a custom logging level
    '''

    logging.addLevelName(102, 'My test level')

    @click.command()
    @click.option('--loglevel', type=click_utils.LogLevelChoice(101, 102))
    def cli(loglevel):
        click.echo('loglevel: {0}'.format(loglevel))

    result = runner.invoke(cli,  ['--help'])

    assert '--loglevel [notset|debug|info|warning|error|critical|level101|mytestlevel]' in result.output

    result = runner.invoke(cli,  ['--loglevel=mytestlevel'], catch_exceptions=False)

    assert result.output == 'loglevel: 102\n'


def test_loglevel_numeric_input(runner):
    '''
    test a special case where user inputted a number loglevel instead of a level name
    '''
    @click.command()
    @click.option('--loglevel', type=click_utils.LogLevelChoice())
    def cli(loglevel):
        assert isinstance(loglevel, int)
        click.echo('loglevel: {0}'.format(loglevel))

    result = runner.invoke(cli,  ['--loglevel=123'], catch_exceptions=False)

    assert result.output == 'loglevel: 123\n'


def test_loglevel_wrong_input(runner):
    '''
    test invokation with wrong input
    '''
    @click.command()
    @click.option('--loglevel', type=click_utils.LogLevelChoice())
    def cli(loglevel):
        click.echo('loglevel: {0}'.format(loglevel))

    result = runner.invoke(cli,  ['--loglevel=nonexistentlevel'], catch_exceptions=True)

    assert result.exit_code != 0
    assert 'Invalid value for "--loglevel": invalid choice: nonexistentlevel' in result.output


def test_loglevel_decorator(runner):
    '''
    test that a shortcut decorator works
    '''
    @click.command()
    @click_utils.loglevel_option()
    def cli(loglevel):
        assert isinstance(loglevel, int)
        click.echo(loglevel)

    result = runner.invoke(cli, ['--loglevel=error'], catch_exceptions=False)

    assert result.exit_code == 0
    assert result.output == '40\n'


def test_loglevel_decorator_custom(runner):
    '''
    test that a shortcut decorator works with a different param name
    '''
    @click.command()
    @click_utils.loglevel_option('--logginglevel')
    def cli(logginglevel):
        assert isinstance(logginglevel, int)
        click.echo(logginglevel)

    result = runner.invoke(cli, ['--logginglevel=info'], catch_exceptions=False)

    assert result.exit_code == 0
    assert result.output == '20\n'


def test_loglevel_uppercase(runner):
    '''
    test invokation with uppercased level name
    '''
    @click.command()
    @click.option('--loglevel', type=click_utils.LogLevelChoice())
    def cli(loglevel):
        click.echo('loglevel: {0}'.format(loglevel))

    result = runner.invoke(cli,  ['--loglevel=WARNING'], catch_exceptions=False)

    assert result.output == 'loglevel: 30\n'


def test_logconfig_option(runner, tmpdir):
    '''
    test logging configuration via logconfig option
    '''
    tmp_file = tmpdir.join('logging.conf')
    conf_content = textwrap.dedent('''
        [loggers]
        keys=root,clickutils_logger

        [handlers]
        keys=hand01

        [formatters]
        keys=form01

        [handler_hand01]
        class=FileHandler
        level=DEBUG
        formatter=form01
        args=('/dev/null',)

        [formatter_form01]
        format=F1 %(asctime)s %(levelname)s %(message)s
        datefmt=
        class=logging.Formatter

        [logger_clickutils_logger]
        level=CRITICAL
        handlers=hand01
        qualname=clickutils_test_logger

        [logger_root]
        handlers=hand01
    ''')

    tmp_file.write(conf_content)

    @click.command()
    @click_utils.logconfig_option()
    def cli():
        pass

    # invoking without configfile shouldn't change anything
    runner.invoke(cli, catch_exceptions=False)
    test_logger = logging.getLogger('clickutils_test_logger')
    assert len(test_logger.handlers) == 0

    # invoking with ``logconfig`` should configure loggers
    runner.invoke(cli,  ['--logconfig={0}'.format(tmp_file.strpath)], catch_exceptions=False)

    test_logger = logging.getLogger('clickutils_test_logger')

    assert len(test_logger.handlers) == 1
    assert test_logger.level == logging.CRITICAL


def test_logfile_option(runner, tmpdir):
    '''
    test a simple case with a logfile option
    '''
    tmp_file = tmpdir.mkdir('logfiles').join('test.log')
    test_logger_name = 'click_utils_test_logfile_option_loger_name'

    @click.command()
    @click_utils.logfile_option(logger_name=test_logger_name)
    def cli():
        pass

    # invoking without logfile shouldn't change anything
    runner.invoke(cli, catch_exceptions=False)
    test_logger = logging.getLogger(test_logger_name)
    assert len(test_logger.handlers) == 0

    # invoking with ``logfile`` should configure a test logger
    runner.invoke(cli,  ['--logfile={0}'.format(tmp_file.strpath)], catch_exceptions=False)
    test_logger = logging.getLogger(test_logger_name)
    assert len(test_logger.handlers) == 1
    assert isinstance(test_logger.handlers[0], logging.handlers.RotatingFileHandler)


def test_logfile_option_with_loglevel_option(runner, tmpdir):
    '''
    test logfile with loglevel and the retrieval of level from context
    '''
    tmp_file = tmpdir.mkdir('logfiles').join('test.log')
    test_logger_name = 'click_utils_test_logfile_option_with_loglevel_option_loger_name'

    @click.command()
    @click_utils.loglevel_option(expose_value=False)
    @click_utils.logfile_option(logger_name=test_logger_name)
    def cli():
        pass

    # invoking without logfile shouldn't change anything
    runner.invoke(cli, catch_exceptions=False)
    test_logger = logging.getLogger(test_logger_name)
    assert len(test_logger.handlers) == 0

    # invoking with ``logfile`` and ``loglevel`` should configure a test logger with a certain level
    runner.invoke(cli,  ['--loglevel=critical', '--logfile={0}'.format(tmp_file.strpath)], catch_exceptions=False)
    test_logger = logging.getLogger(test_logger_name)
    assert len(test_logger.handlers) == 1
    assert isinstance(test_logger.handlers[0], logging.handlers.RotatingFileHandler)

    assert test_logger.level == logging.CRITICAL


def test_logfile_option_with_storing(runner, tmpdir):
    '''
    test that storing a logger on context (from logfile_option) works
    '''
    tmp_file = tmpdir.mkdir('logfiles').join('test.log')
    tmp_file2 = tmpdir.mkdir('logfiles2').join('test2.log')
    test_logger_name = 'click_utils_test_logfile_option_with_storing_loger_name'

    @click.command()
    @click_utils.logfile_option(logger_name=test_logger_name,
                                ctx_key='logger_on_context',
                                default=tmp_file2.strpath)
    @click.pass_context
    def cli(ctx):
        assert isinstance(ctx.logger_on_context.handlers[0], logging.handlers.RotatingFileHandler)

    # invoking without logfile should configure a test logger with default file name and store it on context
    runner.invoke(cli, catch_exceptions=False)
    test_logger = logging.getLogger(test_logger_name)
    assert len(test_logger.handlers) == 1
    assert isinstance(test_logger.handlers[0], logging.handlers.RotatingFileHandler)
    assert test_logger.handlers[0].baseFilename == tmp_file2.strpath

    # invoking with ``logfile`` should configure a test logger with passed file name and store it on context
    runner.invoke(cli,  ['--logfile={0}'.format(tmp_file.strpath)], catch_exceptions=False)
    test_logger = logging.getLogger(test_logger_name)
    assert len(test_logger.handlers) == 2
    assert isinstance(test_logger.handlers[1], logging.handlers.RotatingFileHandler)
    assert test_logger.handlers[1].baseFilename == tmp_file.strpath


def test_logfile_option_with_filters(runner, tmpdir):
    '''
    test that passing an iterable of filters to logfile_option works
    '''
    tmp_file = tmpdir.mkdir('logfiles').join('test.log')
    test_logger_name = 'click_utils_test_logfile_option_with_filters_loger_name'

    class CustomFilter(logging.Filter):
        pass

    filter = CustomFilter()

    @click.command()
    @click_utils.logfile_option(logger_name=test_logger_name, filters=(filter,))
    def cli():
        pass

    runner.invoke(cli,  ['--logfile={0}'.format(tmp_file.strpath)], catch_exceptions=False)
    test_logger = logging.getLogger(test_logger_name)
    assert len(test_logger.filters) == 1

    assert isinstance(test_logger.filters[0], CustomFilter)


def test_default_int_loglevel(runner):
    '''
    test that a loglevel option works when a default value is int
    '''
    @click.command()
    @click_utils.loglevel_option(default=logging.WARNING)
    def cli(loglevel):
        assert isinstance(loglevel, int)
        click.echo(loglevel)

    result = runner.invoke(cli, catch_exceptions=False)
    assert result.exit_code == 0
    assert result.output == '30\n'

    result = runner.invoke(cli, ['--loglevel=error'], catch_exceptions=False)

    assert result.exit_code == 0
    assert result.output == '40\n'


click_utils/help.py
# -*- coding: utf-8 -*-

import click


class EnvHelpOption(click.Option):
    '''
    '''

    def get_help_record(self, ctx):
        ret = super(EnvHelpOption, self).get_help_record(ctx)
        envvars = self.get_envvar_names(ctx)
        if envvars:
            prefix = ' ' if ret[1] else ''
            return ret[0], ret[1] + prefix + '[env: %s]' % ', '.join(envvars)
        return ret

    def get_envvar_names(self, ctx):
        if self.envvar is not None:
            if isinstance(self.envvar, (list, tuple)):
                return self.envvar
            else:
                return [self.envvar]
        else:
            if self.allow_from_autoenv and ctx.auto_envvar_prefix is not None:
                envvar = '%s_%s' % (ctx.auto_envvar_prefix, self.name.upper())
                return [envvar]


class EnvHelpCommand(click.Command):
    '''
    '''

    def __init__(self, *args, **kwargs):

        patch_options = kwargs.pop('envhelp_patch_options', True)
        # patch_subcommands = kwargs.pop('envhelp_patch_subcommands', False)

        super(EnvHelpCommand, self).__init__(*args, **kwargs)

        if patch_options:
            for param in self.params:
                if isinstance(param, click.Option):
                    param.__class__ = EnvHelpOption

        # if patch_subcommands:
        #     pass


click_utils/types.py
# -*- coding: utf-8 -*-

import logging

import click


class LogLevelChoice(click.Choice):
    '''
    A subclass of ``click.Choice`` class for specifying a logging level.

    Example usage::

        import click
        import click_utils

        @click.command()
        @click.option('--loglevel', type=click_utils.LogLevelChoice())
        def cli(loglevel):
            click.echo(loglevel)

    This option type returns an integer representation of a logging level::

        $ cli --loglevel=error
        40

    By default available choices are the lowercased standard ``logging`` level names
    (i.e ``notset``, ``debug``, ``info``, ``warning``, ``error`` and ``critical``)::

        $ cli --help

        Usage: cli [OPTIONS]

        Options:
          --loglevel [notset|debug|info|warning|error|critical]
        ...

    But you can pass your additional logging levels like this::

        logging.addLevelName(102, 'My custom level')

        @click.command()
        @click.option('--loglevel', type=click_utils.LogLevelChoice(101, 102))
        def cli(loglevel):
            click.echo(loglevel)

        ...

        $ cli --help

        Usage: cli [OPTIONS]

        Options:
          --loglevel [notset|debug|info|warning|error|critical|level101|mycustomlevel]
        ...

        $ cli --loglevel=level101
        101

    Finally, as a special case, user can provide any integer as a logging level::

        $ cli --loglevel=123
        123

    '''

    LOGGING_LEVELS = (
        ('notset',   logging.NOTSET),
        ('debug',    logging.DEBUG),
        ('info',     logging.INFO),
        ('warning',  logging.WARNING),
        ('error',    logging.ERROR),
        ('critical', logging.CRITICAL),
    )

    @staticmethod
    def _convert_key(key):
        return key.lower().replace(' ', '')

    def __init__(self, *extra_levels):
        self.logging_levels_dict = dict(self.LOGGING_LEVELS)
        self.logging_level_keys = list(map(lambda x: x[0], self.LOGGING_LEVELS))
        for extra_level in extra_levels:
            level_name = logging.getLevelName(extra_level)
            key = self._convert_key(level_name)
            self.logging_levels_dict.update({key: extra_level})
            self.logging_level_keys.append(key)

        super(LogLevelChoice, self).__init__(self.logging_level_keys)

    def convert(self, value, param, ctx):
        lower_val = self._convert_key(value)
        if lower_val.isdigit():
            return int(lower_val)

        result = super(LogLevelChoice, self).convert(lower_val, param, ctx)
        return self.logging_levels_dict[result]


click_utils/logging.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import

import logging
import logging.config

import click

from . import defaults


class LogLevelChoice(click.Choice):
    '''
    A subclass of ``click.Choice`` class for specifying a logging level.

    Example usage::

        import click
        import click_utils

        @click.command()
        @click.option('--loglevel', type=click_utils.LogLevelChoice())
        def cli(loglevel):
            click.echo(loglevel)

    This option type returns an integer representation of a logging level::

        $ cli --loglevel=error
        40

    By default available choices are the lowercased standard ``logging`` level names
    (i.e ``notset``, ``debug``, ``info``, ``warning``, ``error`` and ``critical``)::

        $ cli --help

        Usage: cli [OPTIONS]

        Options:
          --loglevel [notset|debug|info|warning|error|critical]
        ...

    But you can pass your additional logging levels like this::

        logging.addLevelName(102, 'My custom level')

        @click.command()
        @click.option('--loglevel', type=click_utils.LogLevelChoice(101, 102))
        def cli(loglevel):
            click.echo(loglevel)

        ...

        $ cli --help

        Usage: cli [OPTIONS]

        Options:
          --loglevel [notset|debug|info|warning|error|critical|level101|mycustomlevel]
        ...

        $ cli --loglevel=level101
        101

    You can also pass level name in the uppercased form::

        $ cli --loglevel=WARNING
        30

    Finally, as a special case, user can provide any integer as a logging level::

        $ cli --loglevel=123
        123

    '''

    LOGGING_LEVELS = (
        ('notset',   logging.NOTSET),
        ('debug',    logging.DEBUG),
        ('info',     logging.INFO),
        ('warning',  logging.WARNING),
        ('error',    logging.ERROR),
        ('critical', logging.CRITICAL),
    )

    @staticmethod
    def _convert_key(key):
        return key.lower().replace(' ', '')

    def __init__(self, *extra_levels):
        self.logging_levels_dict = dict(self.LOGGING_LEVELS)
        self.logging_level_keys = list(map(lambda x: x[0], self.LOGGING_LEVELS))
        for extra_level in extra_levels:
            level_name = logging.getLevelName(extra_level)
            key = self._convert_key(level_name)
            self.logging_levels_dict.update({key: extra_level})
            self.logging_level_keys.append(key)

        super(LogLevelChoice, self).__init__(self.logging_level_keys)

    def convert(self, value, param, ctx):
        lower_val = self._convert_key(value) if not isinstance(value, int) else '{0}'.format(value)
        if lower_val.isdigit():
            return int(lower_val)

        result = super(LogLevelChoice, self).convert(lower_val, param, ctx)
        return self.logging_levels_dict[result]


def loglevel_callback_factory(ctx_loglevel_attr=None):

    def inner(ctx, param, value):

        if not value or ctx.resilient_parsing:
            return
        if ctx_loglevel_attr:
            setattr(ctx, ctx_loglevel_attr, value)

        return value
    return inner


def loglevel_option(*param_decls, **attrs):
    '''Shortcut for logging level option type.

    This is equivalent to decorating a function with :func:`option` with
    the following parameters::

        @click.command()
        @click.option('--loglevel', type=click_utils.LogLevelChoice())
        def cli(loglevel):
            pass

    It also has a callback that stores a chosen loglevel on the context object.
    By default the loglevel is stored in attribute called ``click_utils_loglevel``

    To change the attribute name you can pass an keyword arg to this option
    called ``ctx_loglevel_attr``. This is useful for other options to know what
    loglevel was set. If you want to disable storing the loglevel on context
    just pass::

        @click.command()
        @click_utils.loglevel_option(ctx_loglevel_attr=False)
        def cli(loglevel):
            click.echo(loglevel)

    Also this option is eager by default.
    '''
    def decorator(f):
        ctx_loglevel_attr = attrs.pop('ctx_loglevel_attr', defaults.CONTEXT_LOGLEVEL_ATTR)
        default_callback = loglevel_callback_factory(ctx_loglevel_attr=ctx_loglevel_attr)
        attrs.setdefault('callback', default_callback)
        attrs.setdefault('is_eager', True)

        attrs.setdefault('type', LogLevelChoice())
        return click.option(*(param_decls or ('--loglevel',)), **attrs)(f)
    return decorator


def logconfig_callback_factory(defaults=None, disable_existing_loggers=False):
    '''
    a factory for creating parametrized callbacks to invoke ``logging.config.fileConfig``
    '''
    def inner(ctx, param, value):
        if not value or ctx.resilient_parsing:
            return
        logging.config.fileConfig(value,
                                  defaults=defaults,
                                  disable_existing_loggers=disable_existing_loggers)
        return value
    return inner

logconfig_callback = logconfig_callback_factory()

logconfig_callback.__doc__ = '''
a default callback for invoking::

    logging.config.fileConfig(value, defaults=None, disable_existing_loggers=False)
'''


def logconfig_option(*param_decls, **attrs):
    '''An option to easily configure logging via ``logging.config.fileConfig``.
    This one allows to specify a file that will be passed as an argument
    to ``logging.config.fileConfig`` function. Using this option like this::

        @click.command()
        @click_utils.logconfig_option()
        def cli():
            pass

    is equivalent to:

    .. code-block:: python

        def mycallback(ctx, param, value):
            if not value or ctx.resilient_parsing:
                return
            logging.config.fileConfig(value,
                                      defaults=None,
                                      disable_existing_loggers=False)

        @click.command()
        @click.option('--logconfig',
                      expose_value=False,
                      type=click.Path(exists=True, file_okay=True, dir_okay=False,
                                      writable=False, readable=True, resolve_path=True)
                      callback=mycallback
                      )
        def cli():
            pass

    This option accepts all the usual arguments and keyword arguments as ``click.option``
    (be careful with passing a different ``callback`` though).
    Additionally it accepts two extra keyword arguments which are passed
    to ``logging.config.fileConfig``:

    * ``fileconfig_defaults`` is passed as ``defaults`` argument (default: ``None``)
    * ``disable_existing_loggers`` is passed as ``disable_existing_loggers`` argument
      (default: ``False``)

    So you can add logconfig option like this::

        @click.command()
        @click_utils.logconfig_option(disable_existing_loggers=True)
        def cli():
            pass
    '''
    def decorator(f):
        path_type = click.Path(exists=True, file_okay=True, dir_okay=False,
                               writable=False, readable=True, resolve_path=True)
        attrs.setdefault('type', path_type)
        attrs.setdefault('expose_value', False)
        disable_existing_loggers = attrs.pop('disable_existing_loggers', False)
        fileconfig_defaults = attrs.pop('fileconfig_defaults', None)

        callback_kwargs = dict(defaults=fileconfig_defaults, disable_existing_loggers=disable_existing_loggers)
        attrs.setdefault('callback', logconfig_callback_factory(**callback_kwargs))

        return click.option(*(param_decls or ('--logconfig',)), **attrs)(f)
    return decorator


def logger_callback_factory(logger_name='',
                            handler_cls=None, handler_attrs=None,
                            formatter_cls=None, formatter_attrs=None,
                            filters=None,
                            level=None, ctx_loglevel_attr=None,
                            ctx_key=None):

    handler_attrs = handler_attrs if handler_attrs else {}
    formatter_attrs = formatter_attrs if formatter_attrs else {}
    filters = filters if filters else ()
    given_level = level
    ctx_loglevel_attr = ctx_loglevel_attr if ctx_loglevel_attr else defaults.CONTEXT_LOGLEVEL_ATTR

    def inner(ctx, param, value):

        if not value or ctx.resilient_parsing:
            return

        handler = handler_cls(value, **handler_attrs)

        if formatter_cls is not None:
            formatter = formatter_cls(**formatter_attrs)
            handler.setFormatter(formatter)

        level = given_level
        if level is None:
            level = getattr(ctx, ctx_loglevel_attr, defaults.LOGLEVEL)

        handler.setLevel(level)

        logger = logging.getLogger(logger_name)
        logger.addHandler(handler)
        logger.setLevel(level)

        for filter_function in filters:
            logger.addFilter(filter_function)

        # save it in context if ``ctx_key`` was passed
        if ctx_key:
            setattr(ctx, ctx_key, logger)

        return value
    return inner


def logger_option(*param_decls, **attrs):
    '''
    An abstract option for configuring a specific handler to a given logger (root logger by default).
    This logger is then attached to a context for further usage. This option accepts
    all the

    :param handler_cls: **required** a logging Handler class. A value of this option
                        will be passed as a first positional argument while creating
                        an instance of the ``handler_cls``.

    :param handler_attrs: a dictionary of keyword arguments passed to ``handler_cls`` while instantiating
    :param formatter_cls: a Formatter class which instance will be added to a handler
    :param formatter_attrs: a dictionary of keyword arguments passed to ``formatter_cls`` while instantiating
    :param filters: an iterable of filters to add to the logger
    :param level: enforce a minimum logging level on a logger and handler. If this is ``None``
                  then a callback will try to retrieve the level from context (e.g. that was
                  stored by ``loglevel_option``). If it fails to find it a default level is ``logging.WARNING``
    :param ctx_loglevel_attr: an attribute name of the context to **find** a stored logging level
                              (by ``loglevel_option`` for example)
    :param ctx_key: an attribute name of the context to **store** a logger.
                    Pass a *falsy* value to disable storing (default: ``None``).
    :param logger_name: a name of a logger to configure (default: ``''`` i.e. a root logger)
    '''
    def decorator(f):
        callback_kwargs = {
            'handler_cls'      : attrs.pop('handler_cls'),
            'handler_attrs'    : attrs.pop('handler_attrs', {}),
            'formatter_cls'    : attrs.pop('formatter_cls', None),
            'formatter_attrs'  : attrs.pop('formatter_attrs', {}),
            'filters'          : attrs.pop('filters', None),
            'level'            : attrs.pop('level', None),
            'ctx_loglevel_attr': attrs.pop('ctx_loglevel_attr', None),
            'ctx_key'          : attrs.pop('ctx_key', None),
            'logger_name'      : attrs.pop('logger_name', ''),
        }

        default_callback = logger_callback_factory(**callback_kwargs)
        attrs.setdefault('callback', default_callback)
        attrs.setdefault('expose_value', False)
        return click.option(*(param_decls or ('--logger',)), **attrs)(f)
    return decorator


def logfile_option(*param_decls, **attrs):
    '''
    a specific type of ``logger_option`` that configures
    a logging file handler (by default it's a ``logging.handlers.RotatingFileHandler``)
    on a logger (root logger by default).

    In addition to all ``click_utils.logger_option`` and ``click.option`` arguments
    it accepts four keyword arguments to control handler creation and format settings.

    :param fmt: a format (``fmt``) argument for ``logging.Formatter`` class
    :param datefmt: a date format (``datefmt``) argument for ``logging.Formatter`` class

    :param maxBytes: a ``maxBytes`` argument for ``RotatingFileHandler`` class
    :param backupCount: a ``backupCount`` argument for ``RotatingFileHandler`` class
    '''

    def decorator(f):
        path_type = click.Path(exists=False, file_okay=True, dir_okay=False,
                               writable=True, readable=True, resolve_path=True)
        attrs.setdefault('type', path_type)

        fmt = attrs.pop('fmt', defaults.FORMAT)
        datefmt = attrs.pop('datefmt', defaults.DATEFMT)

        maxBytes = attrs.pop('maxBytes', defaults.MAXBYTES)
        backupCount = attrs.pop('backupCount', defaults.BACKUPCOUNT)

        # logger_option attrs
        attrs.setdefault('handler_cls', logging.handlers.RotatingFileHandler),
        attrs.setdefault('handler_attrs', {'maxBytes': maxBytes, 'backupCount': backupCount}),
        attrs.setdefault('formatter_cls', logging.Formatter),
        attrs.setdefault('formatter_attrs', {'fmt': fmt, 'datefmt': datefmt}),
        attrs.setdefault('filters', None),
        attrs.setdefault('level', None),
        attrs.setdefault('ctx_loglevel_attr', None),
        attrs.setdefault('ctx_key', None),
        attrs.setdefault('logger_name', ''),

        return logger_option(*(param_decls or ('--logfile',)), **attrs)(f)
    return decorator
